File: samples/CslaLight/vb/ADONetDataServicesSample/DataServices.Business.Server/BusinessClasses/Company.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports Csla
Imports Csla.Security
Imports Csla.Core
Imports Csla.Serialization
Imports Csla.DataPortalClient
Imports System.ComponentModel
Imports Csla.Validation
Imports System.Data

#If Not SILVERLIGHT = 1 Then
Imports System.Data.SqlClient
#End If

<Serializable()> _
Public Class Company
  Inherits BusinessBase(Of Company)


#If Silverlight = 1 Then
  Public Sub New()
  End Sub
#Else
  Private Sub New()
  End Sub
#End If

  Public Shared Function GetServiceUri() As Uri
    Return New Uri("CompanyService.svc", UriKind.Relative)
  End Function


  Private Shared CompanyIdProperty As PropertyInfo(Of Integer) = RegisterProperty(Of Integer)(New PropertyInfo(Of Integer)("CompanyId", "Company Id", 0))
  Public ReadOnly Property CompanyId() As Integer
    Get
      Return GetProperty(CompanyIdProperty)
    End Get
  End Property

  Private Shared CompanyNameProperty As PropertyInfo(Of String) = RegisterProperty(Of String)(New PropertyInfo(Of String)("CompanyName", "Company Name", String.Empty))
  Public Property CompanyName() As String
    Get
      Return GetProperty(CompanyNameProperty)
    End Get
    Set(ByVal value As String)
      SetProperty(CompanyNameProperty, value)
    End Set
  End Property

  Private Shared DateAddedProperty As PropertyInfo(Of SmartDate) = RegisterProperty(Of SmartDate)(New PropertyInfo(Of SmartDate)("DateAdded", "Date Added"))
  Public Property DateAdded() As String
    Get
      Return GetProperty(DateAddedProperty).Text
    End Get
    Set(ByVal value As String)
      Dim test As SmartDate = New SmartDate()
      If SmartDate.TryParse(value, test) Then
        SetProperty(DateAddedProperty, test)
      End If
    End Set
  End Property

  Friend ReadOnly Property DateAddedValue() As Object
    Get
      Return GetProperty(DateAddedProperty).DBValue
    End Get
  End Property

  Protected Overrides Sub AddAuthorizationRules()
    MyBase.AddAuthorizationRules()
    Dim canWrite() As String = New String() {"AdminUser", "RegularUser"}
    Dim canRead() As String = New String() {"AdminUser", "RegularUser", "ReadOnlyUser"}
    Dim admin() As String = New String() {"AdminUser"}
    AuthorizationRules.AllowCreate(GetType(Company), admin)
    AuthorizationRules.AllowDelete(GetType(Company), admin)
    AuthorizationRules.AllowEdit(GetType(Company), canWrite)
    AuthorizationRules.AllowGet(GetType(Company), canRead)
    AuthorizationRules.AllowWrite(CompanyNameProperty, canWrite)
    AuthorizationRules.AllowWrite(DateAddedProperty, canWrite)
    AuthorizationRules.AllowRead(CompanyNameProperty, canRead)
    AuthorizationRules.AllowRead(CompanyIdProperty, canRead)
    AuthorizationRules.AllowRead(DateAddedProperty, canRead)
  End Sub

  Protected Overrides Sub AddBusinessRules()
    MyBase.AddBusinessRules()
    ValidationRules.AddRule(AddressOf CommonRules.StringRequired, New RuleArgs(CompanyNameProperty))
    ValidationRules.AddRule(AddressOf CommonRules.StringMaxLength, New CommonRules.MaxLengthRuleArgs(CompanyNameProperty, 50))
    ValidationRules.AddRule(Of Company)(AddressOf IsDateValid, DateAddedProperty)

  End Sub

  Private Shared Function IsDateValid(ByVal target As Company, ByVal e As RuleArgs) As Boolean
    Dim dateAdded As SmartDate = target.GetProperty(DateAddedProperty)
    If Not dateAdded.IsEmpty Then
      If (dateAdded.Date < New DateTime(2000, 1, 1)) Then
        e.Description = "Date must be greater that 1/1/2000!"
        Return False
      ElseIf (dateAdded.Date > DateTime.Today) Then
        e.Description = "Date cannot be greater than today!"
        Return False
      End If
    Else
      e.Description = "Date added is required!"
      Return False
    End If
    Return True
  End Function

  Friend Sub SetID(ByVal companyId As Integer)
    LoadProperty(CompanyIdProperty, companyId)
  End Sub

  Public Shared Sub GetCompany(ByVal companyId As Integer, ByVal handler As EventHandler(Of DataPortalResult(Of Company)))
    Dim dp As New DataPortal(Of Company)()
    AddHandler dp.FetchCompleted, handler
    dp.BeginFetch(New SingleCriteria(Of Company, Integer)(companyId))
  End Sub

  Public Shared Sub CreateCompany(ByVal handler As EventHandler(Of DataPortalResult(Of Company)))
    Dim dp As New DataPortal(Of Company)()
    AddHandler dp.CreateCompleted, handler
    dp.BeginCreate()
  End Sub

#If Silverlight = 1 Then

  Private _handler As LocalProxy(Of Company).CompletedHandler
  Private _criteria As SingleCriteria(Of Company, Integer)

  <EditorBrowsable(EditorBrowsableState.Never)> _
  Public Sub DataPortal_Fetch(ByVal criteria As SingleCriteria(Of Company, Integer), ByVal handler As LocalProxy(Of Company).CompletedHandler)
    Try
      _handler = handler
      _criteria = criteria
      Dim context As CompanyServiceReference.CompanyEntities = Csla.Data.DataServiceContextManager(Of CompanyServiceReference.CompanyEntities).GetManager(GetServiceUri).DataServiceContext

      Dim query = From oneCompany In context.Companies _
                  Where oneCompany.CompanyId = criteria.Value _
                  Select oneCompany
      Dim companyQuery As Services.Client.DataServiceQuery(Of CompanyServiceReference.Companies) = CType(query, Global.System.Data.Services.Client.DataServiceQuery(Of Global.DataServices.Business.CompanyServiceReference.Companies))
      companyQuery.BeginExecute(AddressOf EndFetch, companyQuery)
    Catch ex As Exception
      _handler(Nothing, ex)
      _handler = Nothing
      _criteria = Nothing
    End Try

  End Sub

  Private Sub EndFetch(ByVal e As IAsyncResult)
    Try
      Dim queryCompany = CType(e.AsyncState, Services.Client.DataServiceQuery(Of CompanyServiceReference.Companies))
      If queryCompany IsNot Nothing Then
        Dim aCompany = queryCompany.EndExecute(e).FirstOrDefault()
        If aCompany IsNot Nothing Then
          LoadProperty(CompanyIdProperty, aCompany.CompanyId)
          LoadProperty(CompanyNameProperty, aCompany.CompanyName)
          LoadProperty(DateAddedProperty, New SmartDate(aCompany.DateAdded))
          _handler(Me, Nothing)
        Else
          _handler(Nothing, New ArgumentException("Company with Id of " & _criteria.Value.ToString() & " is not found."))
        End If
      Else
        _handler(Nothing, New ArgumentException("Company with Id of " & _criteria.Value.ToString() & " is not found."))
      End If
    Catch ex As Exception
      _handler(Nothing, ex)
    Finally
      _handler = Nothing
      _criteria = Nothing
    End Try

  End Sub

  <EditorBrowsable(EditorBrowsableState.Never)> _
  Public Shadows Sub DataPortal_DeleteSelf(ByVal handler As LocalProxy(Of Company).CompletedHandler)
    If Not IsNew Then
      _handler = handler
      Dim context As CompanyServiceReference.CompanyEntities = Csla.Data.DataServiceContextManager(Of CompanyServiceReference.CompanyEntities).GetManager(GetServiceUri).DataServiceContext
      Dim company As CompanyServiceReference.Companies = GetCompany()
      context.DeleteObject(company)
      context.BeginSaveChanges(Services.Client.SaveChangesOptions.Batch, AddressOf EndUpdate, Nothing)
    Else
      _handler(Me, Nothing)
    End If

  End Sub

  <EditorBrowsable(EditorBrowsableState.Never)> _
  Public Shadows Sub DataPortal_Insert(ByVal handler As LocalProxy(Of Company).CompletedHandler)
    Try
      _handler = handler
      Dim context As CompanyServiceReference.CompanyEntities = Csla.Data.DataServiceContextManager(Of CompanyServiceReference.CompanyEntities).GetManager(GetServiceUri).DataServiceContext
      Dim company As CompanyServiceReference.Companies = New CompanyServiceReference.Companies
      With company
        .CompanyId = ReadProperty(CompanyIdProperty)
        .CompanyName = ReadProperty(CompanyNameProperty)
        .DateAdded = ReadProperty(DateAddedProperty)
      End With
      context.AddToCompanies(company)
      context.BeginSaveChanges(Services.Client.SaveChangesOptions.Batch, AddressOf EndUpdate, Nothing)
    Catch ex As Exception
      _handler(Me, ex)
    End Try

  End Sub

  <EditorBrowsable(EditorBrowsableState.Never)> _
  Public Shadows Sub DataPortal_Update(ByVal handler As LocalProxy(Of Company).CompletedHandler)

    Try
      If IsDirty Then
        _handler = handler
        Dim context As CompanyServiceReference.CompanyEntities = Csla.Data.DataServiceContextManager(Of CompanyServiceReference.CompanyEntities).GetManager(GetServiceUri).DataServiceContext
        Dim company As CompanyServiceReference.Companies = GetCompany()
        With company
          .CompanyId = ReadProperty(CompanyIdProperty)
          .CompanyName = ReadProperty(CompanyNameProperty)
          .DateAdded = ReadProperty(DateAddedProperty)
        End With
        context.UpdateObject(company)
        context.BeginSaveChanges(Services.Client.SaveChangesOptions.Batch, AddressOf EndUpdate, Nothing)
      Else
        _handler(Me, Nothing)
      End If
    Catch ex As Exception
      _handler(Me, ex)
    End Try

  End Sub

  Private Sub EndUpdate(ByVal e As IAsyncResult)
    Try
      Dim context As CompanyServiceReference.CompanyEntities = Csla.Data.DataServiceContextManager(Of CompanyServiceReference.CompanyEntities).GetManager(GetServiceUri).DataServiceContext
      Dim response As Services.Client.DataServiceResponse = context.EndSaveChanges(e)
      Dim anError As Exception = Nothing
      For Each oneResponse As Services.Client.OperationResponse In response
        If oneResponse.Error IsNot Nothing Then
          anError = oneResponse.Error
        Else
          If Not IsDeleted Then
            Dim details As System.Data.Services.Client.ChangeOperationResponse = CType(oneResponse, Services.Client.ChangeOperationResponse)
            Dim company As CompanyServiceReference.Companies = CType(CType(details.Descriptor, Services.Client.EntityDescriptor).Entity, CompanyServiceReference.Companies)
            LoadProperty(CompanyIdProperty, company.CompanyId)
          End If
        End If
      Next
      If anError Is Nothing Then
        If Not IsDeleted Then
          MarkOld()
        End If
        _handler(Me, Nothing)
        _handler = Nothing
      Else
        _handler(Me, anError)
      End If
    Catch ex As Exception
      _handler(Me, ex)
    End Try

  End Sub

  Private Function GetCompany() As CompanyServiceReference.Companies
    Dim returnValue As CompanyServiceReference.Companies = Nothing
    Dim manager = Csla.Data.DataServiceContextManager(Of CompanyServiceReference.CompanyEntities).GetManager(GetServiceUri)
    returnValue = manager.GetEntity(Of CompanyServiceReference.Companies)(CompanyIdProperty.Name, ReadProperty(CompanyIdProperty))
    Return returnValue
  End Function

#End If

End Class

File: Samples/CslaLight/cs/ADONetDataServicesSample/DataServices.Business.Server/BusinessClasses/Company.cs
//INSTANT C# NOTE: Formerly VB.NET project-level imports:
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using System.Linq;

using System.Text;
using Csla;
using Csla.Security;
using Csla.Core;
using Csla.Serialization;
using Csla.DataPortalClient;
using System.ComponentModel;
using Csla.Validation;

#if !SILVERLIGHT
using System.Data.SqlClient;
#endif

namespace DataServices.Business
{
  [Serializable()]
  public class Company : BusinessBase<Company>
  {


#if SILVERLIGHT
    public Company()
    {
    }
#else
    private Company()
    {
    }
#endif

    public static Uri GetServiceUri()
    {
      return new Uri("CompanyService.svc", UriKind.Relative);
    }


    private static PropertyInfo<int> CompanyIdProperty = RegisterProperty<int>(new PropertyInfo<int>("CompanyId", "Company Id", 0));
    public int CompanyId
    {
      get
      {
        return GetProperty(CompanyIdProperty);
      }
    }

    private static PropertyInfo<string> CompanyNameProperty = RegisterProperty<string>(new PropertyInfo<string>("CompanyName", "Company Name", string.Empty));
    public string CompanyName
    {
      get
      {
        return GetProperty(CompanyNameProperty);
      }
      set
      {
        SetProperty(CompanyNameProperty, value);
      }
    }

    private static PropertyInfo<SmartDate> DateAddedProperty = RegisterProperty<SmartDate>(new PropertyInfo<SmartDate>("DateAdded", "Date Added"));
    public string DateAdded
    {
      get
      {
        return GetProperty(DateAddedProperty).Text;
      }
      set
      {
        SmartDate test = new SmartDate();
        if (SmartDate.TryParse(value, ref test))
        {
          SetProperty(DateAddedProperty, test);
        }
      }
    }

    internal object DateAddedValue
    {
      get
      {
        return GetProperty(DateAddedProperty).DBValue;
      }
    }

    protected override void AddAuthorizationRules()
    {
      base.AddAuthorizationRules();
      string[] canWrite = new string[] { "AdminUser", "RegularUser" };
      string[] canRead = new string[] { "AdminUser", "RegularUser", "ReadOnlyUser" };
      string[] admin = new string[] { "AdminUser" };
      AuthorizationRules.AllowCreate(typeof(Company), admin);
      AuthorizationRules.AllowDelete(typeof(Company), admin);
      AuthorizationRules.AllowEdit(typeof(Company), canWrite);
      AuthorizationRules.AllowGet(typeof(Company), canRead);
      AuthorizationRules.AllowWrite(CompanyNameProperty, canWrite);
      AuthorizationRules.AllowWrite(DateAddedProperty, canWrite);
      AuthorizationRules.AllowRead(CompanyNameProperty, canRead);
      AuthorizationRules.AllowRead(CompanyIdProperty, canRead);
      AuthorizationRules.AllowRead(DateAddedProperty, canRead);
    }

    protected override void AddBusinessRules()
    {
      base.AddBusinessRules();
      ValidationRules.AddRule(CommonRules.StringRequired, new RuleArgs(CompanyNameProperty));
      ValidationRules.AddRule(CommonRules.StringMaxLength, new CommonRules.MaxLengthRuleArgs(CompanyNameProperty, 50));
      ValidationRules.AddRule<Company>(IsDateValid, DateAddedProperty);

    }

    private static bool IsDateValid(Company target, RuleArgs e)
    {
      SmartDate dateAdded = target.GetProperty(DateAddedProperty);
      if (!dateAdded.IsEmpty)
      {
        if (dateAdded.Date < new System.DateTime(2000, 1, 1))
        {
          e.Description = "Date must be greater that 1/1/2000!";
          return false;
        }
        else if (dateAdded.Date > DateTime.Today)
        {
          e.Description = "Date cannot be greater than today!";
          return false;
        }
      }
      else
      {
        e.Description = "Date added is required!";
        return false;
      }
      return true;
    }

    internal void SetID(int companyId)
    {
      LoadProperty(CompanyIdProperty, companyId);
    }

    public static void GetCompany(int companyId, EventHandler<DataPortalResult<Company>> handler)
    {
      DataPortal<Company> dp = new DataPortal<Company>();
      dp.FetchCompleted += handler;
      dp.BeginFetch(new SingleCriteria<Company, int>(companyId));
    }

    public static void CreateCompany(EventHandler<DataPortalResult<Company>> handler)
    {
      DataPortal<Company> dp = new DataPortal<Company>();
      dp.CreateCompleted += handler;
      dp.BeginCreate();
    }

#if SILVERLIGHT

    private LocalProxy<Company>.CompletedHandler _handler;
    private SingleCriteria<Company, int> _criteria;

    [EditorBrowsable(EditorBrowsableState.Never)]
    public void DataPortal_Fetch(SingleCriteria<Company, int> criteria, LocalProxy<Company>.CompletedHandler handler)
    {
      try
      {
        _handler = handler;
        _criteria = criteria;
        CompanyServiceReference.CompanyEntities context = Csla.Data.DataServiceContextManager<CompanyServiceReference.CompanyEntities>.GetManager(GetServiceUri()).DataServiceContext;

        var query =
          from oneCompany in context.Companies
          where oneCompany.CompanyId == criteria.Value
          select oneCompany;
        global::System.Data.Services.Client.DataServiceQuery<CompanyServiceReference.Companies> companyQuery = (global::System.Data.Services.Client.DataServiceQuery<global::DataServices.Business.CompanyServiceReference.Companies>)query;
        companyQuery.BeginExecute(EndFetch, companyQuery);
      }
      catch (Exception ex)
      {
        _handler(null, ex);
        _handler = null;
        _criteria = null;
      }

    }

    private void EndFetch(IAsyncResult e)
    {
      try
      {
        var queryCompany = (System.Data.Services.Client.DataServiceQuery<CompanyServiceReference.Companies>)e.AsyncState;
        if (queryCompany != null)
        {
          var aCompany = queryCompany.EndExecute(e).FirstOrDefault();
          if (aCompany != null)
          {
            LoadProperty(CompanyIdProperty, aCompany.CompanyId);
            LoadProperty(CompanyNameProperty, aCompany.CompanyName);
            LoadProperty(DateAddedProperty, new SmartDate(aCompany.DateAdded));
            _handler(this, null);
          }
          else
          {
            _handler(null, new ArgumentException("Company with Id of " + _criteria.Value.ToString() + " is not found."));
          }
        }
        else
        {
          _handler(null, new ArgumentException("Company with Id of " + _criteria.Value.ToString() + " is not found."));
        }
      }
      catch (Exception ex)
      {
        _handler(null, ex);
      }
      finally
      {
        _handler = null;
        _criteria = null;
      }

    }

    [EditorBrowsable(EditorBrowsableState.Never)]
    public new void DataPortal_DeleteSelf(LocalProxy<Company>.CompletedHandler handler)
    {
      if (!IsNew)
      {
        _handler = handler;
        CompanyServiceReference.CompanyEntities context = Csla.Data.DataServiceContextManager<CompanyServiceReference.CompanyEntities>.GetManager(GetServiceUri()).DataServiceContext;
        CompanyServiceReference.Companies company = GetCompany();
        context.DeleteObject(company);
        context.BeginSaveChanges(System.Data.Services.Client.SaveChangesOptions.Batch, EndUpdate, null);
      }
      else
      {
        _handler(this, null);
      }

    }

    [EditorBrowsable(EditorBrowsableState.Never)]
    public new void DataPortal_Insert(LocalProxy<Company>.CompletedHandler handler)
    {
      try
      {
        _handler = handler;
        CompanyServiceReference.CompanyEntities context = Csla.Data.DataServiceContextManager<CompanyServiceReference.CompanyEntities>.GetManager(GetServiceUri()).DataServiceContext;
        CompanyServiceReference.Companies company = new CompanyServiceReference.Companies();
        company.CompanyId = ReadProperty(CompanyIdProperty);
        company.CompanyName = ReadProperty(CompanyNameProperty);
        company.DateAdded = ReadProperty(DateAddedProperty);
        context.AddToCompanies(company);
        context.BeginSaveChanges(System.Data.Services.Client.SaveChangesOptions.Batch, EndUpdate, null);
      }
      catch (Exception ex)
      {
        _handler(this, ex);
      }

    }

    [EditorBrowsable(EditorBrowsableState.Never)]
    public new void DataPortal_Update(LocalProxy<Company>.CompletedHandler handler)
    {

      try
      {
        if (IsDirty)
        {
          _handler = handler;
          CompanyServiceReference.CompanyEntities context = Csla.Data.DataServiceContextManager<CompanyServiceReference.CompanyEntities>.GetManager(GetServiceUri()).DataServiceContext;
          CompanyServiceReference.Companies company = GetCompany();
          company.CompanyId = ReadProperty(CompanyIdProperty);
          company.CompanyName = ReadProperty(CompanyNameProperty);
          company.DateAdded = ReadProperty(DateAddedProperty);
          context.UpdateObject(company);
          context.BeginSaveChanges(System.Data.Services.Client.SaveChangesOptions.Batch, EndUpdate, null);
        }
        else
        {
          _handler(this, null);
        }
      }
      catch (Exception ex)
      {
        _handler(this, ex);
      }

    }

    private void EndUpdate(IAsyncResult e)
    {
      try
      {
        CompanyServiceReference.CompanyEntities context = Csla.Data.DataServiceContextManager<CompanyServiceReference.CompanyEntities>.GetManager(GetServiceUri()).DataServiceContext;
        System.Data.Services.Client.DataServiceResponse response = context.EndSaveChanges(e);
        Exception anError = null;
        foreach (System.Data.Services.Client.OperationResponse oneResponse in response)
        {
          if (oneResponse.Error != null)
          {
            anError = oneResponse.Error;
          }
          else
          {
            if (!IsDeleted)
            {
              System.Data.Services.Client.ChangeOperationResponse details = (System.Data.Services.Client.ChangeOperationResponse)oneResponse;
              CompanyServiceReference.Companies company = (CompanyServiceReference.Companies)(((System.Data.Services.Client.EntityDescriptor)details.Descriptor).Entity);
              LoadProperty(CompanyIdProperty, company.CompanyId);
            }
          }
        }
        if (anError == null)
        {
          if (!IsDeleted)
          {
            MarkOld();
          }
          _handler(this, null);
          _handler = null;
        }
        else
        {
          _handler(this, anError);
        }
      }
      catch (Exception ex)
      {
        _handler(this, ex);
      }

    }

    private CompanyServiceReference.Companies GetCompany()
    {
      CompanyServiceReference.Companies returnValue = null;
      var manager = Csla.Data.DataServiceContextManager<CompanyServiceReference.CompanyEntities>.GetManager(GetServiceUri());
      returnValue = manager.GetEntity<CompanyServiceReference.Companies>(CompanyIdProperty.Name, ReadProperty(CompanyIdProperty));
      return returnValue;
    }

#endif

  }

} //end of root namespace